Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,600 @@
+            }
+
+            connection.whenFailure { _ in
+                let action = self.stateLock.withLock {
+                    self.state.failedConnectionAction()
+                }
+                switch action {
+                case .makeConnectionAndComplete(let el, let promise):
+                    self.makeConnection(on: el).cascade(to: promise)
+                case .none:
+                    break
+                }
+            }
+
+            return connection
+        }
+
+        /// Adds a callback on connection close that asks the `state` what to do about this
+        ///
+        /// The callback informs the state about the event, and the state returns a
+        /// `ClosedConnectionRemoveAction` which instructs it about what it should do.
+        private func configureCloseCallback(of connection: Connection) {
+            connection.channel.closeFuture.whenComplete { result in
+                switch result {
+                case .success:
+                    let action = self.parentPool.connectionProvidersLock.withLock {
+                        self.stateLock.withLock { self.state.removeClosedConnection(connection) }
+                    }
+                    switch action {
+                    case .none:
+                        break
+                    case .makeConnectionAndComplete(let el, let promise):
+                        self.makeConnection(on: el).cascade(to: promise)
+                    }
+
+                case .failure(let error):
+                    preconditionFailure("Connection close future failed with error: \(error)")
+                }
+            }
+        }
+
+        func syncClose(requiresCleanClose: Bool) throws {
+            var closeError: Error?
+            let availableConnections = self.stateLock.withLock { () -> CircularBuffer<ConnectionPool.Connection> in
+                assert(!self.state.isClosed, "Calling syncClose on an already closed provider")
+                let waitersCopy = self.state.waiters
+                self.state.waiters.removeAll()
+                for waiter in waitersCopy {
+                    waiter.promise.fail(HTTPClientError.cancelled)
+                }
+                self.state.isClosed = true
+                assert(self.state.leased == 0, "Invalid number of leased connections on close: \(self.state.leased)")
+                return self.state.availableConnections
+            }
+            do {
+                try EventLoopFuture<Void>.andAllComplete(availableConnections.map { $0.channel.close() }, on: self.eventLoop).wait()
+            } catch ChannelError.alreadyClosed {
+                return
+            } catch {
+                // First catched errors have priority, this
+                // is to be consistent with what would happen
+                // if the `throw` wasn't delayed until everything
+                // is executed
+                if closeError == nil {
+                    closeError = error
```

Comment:
Is there any other possible error than `ChannelError.alreadyClosed` on close? If there is a reproducible one it might be good to add a test that makes it so that this function throws

Reply:
depends on the channel handlers in the pipeline to be precise. I think ignoring `alreadyClosed` and surfacing everything else makes sense